backend/aethel_db/views.py

```diff
@@ -85,32 +79,65 @@ def get(self, request: HttpRequest) -> JsonResponse:
-        query_result = search(
-            bank=dataset.samples,
-            query=in_word(query_input) | in_lemma(query_input),
-        )
-
-        def item_contains_query_string(item: LexicalItem) -> bool:
+        def item_contains_query_string(item: LexicalItem, query_input: str) -> bool:
+            """
+            Checks if a LexicalItem contains a given input string in its word or its lemma.
+            """
             return (
                 query_input.lower() in item.lemma.lower()
                 or query_input.lower() in item.word.lower()
             )
 
+        def serialize_sample(
+            sample: Sample, highlighted_phrase_indices: set[int]
+        ) -> AethelListSample:
+            """
+            Turns a Sample into an AethelListSample, while marking phrases that should be highlighted.
+            """
+            new_phrases = []
+            for index, phrase in enumerate(sample.lexical_phrases):
+                highlighted = index in highlighted_phrase_indices
+                new_phrase = AethelSamplePhrase(
+                    index=index,
+                    display=phrase.string,
+                    highlight=highlighted,
+                )
+                new_phrases.append(new_phrase)
+
+            return AethelListSample(sample.name, new_phrases)
+
         response_object = AethelListResponse()
-        # Then we loop over the samples and extract what we need from them (lemma, word, type etc.).
+
+        # First we select all relevant samples from the dataset that contain the query string.
+        query_result = search(
+            bank=dataset.samples,
+            query=in_word(query_input) | in_lemma(query_input),
+        )
+
+        # Then we transform the results.
+        # Each key in result_dict is a unique combination of lemma, word, and type.
+        # Each value is a sample, mapped to a set of indices referring to the specific phrase that has the type.
+        result_dict: dict[tuple[str, str, str], dict[str, set]] = {}
         for sample in query_result:
-            lexical_phrases = sample.lexical_phrases
-            for phrase in lexical_phrases:
+            for phrase_index, phrase in enumerate(sample.lexical_phrases):
                 for item in phrase.items:
-                    if item_contains_query_string(item):
-                        response_object.add_result(
-                            item.lemma,
-                            item.word,
-                            str(phrase.type),
-                            sample.name,
-                            sample.sentence,
-                        )
+                    if item_contains_query_string(item, query_input):
+                        key = (item.lemma, item.word, str(phrase.type))
+                        # setdefault gets the value for a given key or adds the key with a provided value if None is found.
+                        samples = result_dict.setdefault(key, {})
```

Comment:
for this kind of stuff I personally like `collections.defaultdict` but this is also ok

Reply:
Ah, I wasn't aware of this. Good to know.